iroh-net/src/lib.rs

```diff
@@ -1,11 +1,102 @@
+//! longer involved in the connection.
+//!
+//! If one of the iroh-net nodes can be reached directly connectivity can also be
+//! established without involving a Relay server.  The [`NodeAddr`] can also contain a
+//! number of [socket addresses] on which a node is reachable and will be used to establish a
+//! connection.
+//!
+//!
+//! # Encryption
+//!
+//! The connection is encrypted using TLS, like standard QUIC connections.  Unlike standard
+//! QUIC there is no client, server or server TLS key.  Instead each iroh-net node has a
+//! unique [`SecretKey`] used to authenticate and encrypt the connection.  When an iroh-net
+//! node connects it uses the corresponding [`PublicKey`] to ensure the connection is only
+//! established with the intended peer.
+//!
+//! Since the [`PublicKey`] is also used to identify the iroh-net node it is also known as
+//! the [`NodeId`].  As encryption is an integral part of TLS as used in QUIC this
+//! [`NodeId`] is always a required parameter to establish a connection.
+//!
+//! When accepting connections the peer's [`NodeId`] is authenticated.  However it is up to
+//! the application to decide if a particular peer is allowed to connect or not.
+//!
+//!
+//! # Relay Servers
+//!
+//! Relay servers exist to ensure all iroh-net nodes are always reachable.  They accept
+//! **encrypted** traffic for iroh-net nodes which are connected to them, forwarding it to
+//! the correct destination based on the [`NodeId`] only.  Since nodes only send encrypted
+//! traffic, the Relay servers can not decode any traffic for other iroh-net nodes and only
+//! forwards it.
+//!
+//! The connections to the Relay server are initiated as normal HTTP 1.1 connections using
+//! TLS.  Once connected the transport is upgraded to a plain TCP connection using a custom
+//! protocol.  All further data is then sent using this custom relaying protocol.  Usually
+//! soon after the connection is established via the Relay it will migrate to a direct
+//! connection.  However if this is not possible the connection will keep flowing over the
+//! relay server as a fallback.
+//!
+//! Additionally to providing reliable connectivity between iroh-net nodes, Relay servers
+//! provide some functions to assist in [hole punching].  They have various services to help
+//! nodes understand their own network situation.  This includes offering a [STUN] server,
+//! but also a few HTTP extra endpoints as well as responding to ICMP echo requests.
+//!
+//!
+//! # Connections and Streams
+//!
+//! Connections are managed using the [`Endpoint`].  To establish a connection to an
+//! iroh-net node you need to know three pieces of information:
+//!
+//! - The [`NodeId`] of the peer to connect to.
+//! - Some addressing information:
+//!   - Usually the [`RelayUrl`] identifying the Relay server.
+//!   - Sometimes, or usually additionally, any direct addresses which might be known.
```

Comment:
Here is worth mentioning dns since we technically might now not need _any_ addressing information beside the node id itself

Reply:
Good catch that I missed this, I added a discovery section below.  Do you think that's a good way of addressing this?